File: workflowtrigger/src/main/java/com/greenops/workflowtrigger/api/PipelineStatusApi.java
package com.greenops.workflowtrigger.api;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.greenops.util.datamodel.auditlog.DeploymentLog;
import com.greenops.util.datamodel.auditlog.Log;
import com.greenops.util.datamodel.pipelinestatus.PipelineStatus;
import com.greenops.util.dbclient.DbClient;
import com.greenops.util.error.AtlasNonRetryableError;
import com.greenops.workflowtrigger.dbclient.DbKey;
import com.greenops.workflowtrigger.validator.RequestSchemaValidator;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.ArrayList;
import java.util.List;

import static com.greenops.util.dbclient.DbClient.LOG_INCREMENT;

@Slf4j
@RestController
@RequestMapping("/status")
public class PipelineStatusApi {

    private final DbClient dbClient;
    private final ObjectMapper objectMapper;
    private final RequestSchemaValidator requestSchemaValidator;

    @Autowired
    public PipelineStatusApi(DbClient dbClient, ObjectMapper objectMapper, RequestSchemaValidator requestSchemaValidator) {
        this.dbClient = dbClient;
        this.objectMapper = objectMapper;
        this.requestSchemaValidator = requestSchemaValidator;
    }

    @GetMapping(value = "{orgName}/{teamName}/pipeline/{pipelineName}/step/{stepName}/{count}")
    public ResponseEntity<List<Log>> getStepLogs(@PathVariable("orgName") String orgName,
                                                 @PathVariable("teamName") String teamName,
                                                 @PathVariable("pipelineName") String pipelineName,
                                                 @PathVariable("stepName") String stepName,
                                                 @PathVariable("count") int count) {
        if (!requestSchemaValidator.checkAuthentication()) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
        }
        var key = DbKey.makeDbStepKey(orgName, teamName, pipelineName, stepName);
        var increments = (int) Math.ceil(LOG_INCREMENT / (double) count);
        var logList = new ArrayList<Log>();
        for (int idx = 0; idx < increments; idx++) {
            var fetchedLogList = dbClient.fetchLogList(key, idx);
            if (idx == increments - 1) {
                var difference = count - ((increments - 1) * LOG_INCREMENT);
                logList.addAll(fetchedLogList.subList(0, Math.min(difference, fetchedLogList.size())));
            } else {
                logList.addAll(fetchedLogList);
            }
        }

        return ResponseEntity.ok(logList);
    }

    @GetMapping(value = "{orgName}/{teamName}/pipeline/{pipelineName}/{pipelineUvn}")
    public ResponseEntity<String> getPipelineStatus(@PathVariable("orgName") String orgName,
                                                    @PathVariable("teamName") String teamName,
                                                    @PathVariable("pipelineName") String pipelineName,
                                                    @PathVariable("pipelineUvn") String pipelineUvn) {
        if (!requestSchemaValidator.checkAuthentication()) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
        }
        var status = new PipelineStatus();
        var steps = dbClient.fetchStringList(DbKey.makeDbListOfStepsKey(orgName, teamName, pipelineName));
        if (steps == null) return ResponseEntity.ok("");
        for (var step : steps) {
            try {
                //Get pipeline UVN if not specified
                var logKey = DbKey.makeDbStepKey(orgName, teamName, pipelineName, step);
                Log log = null;
                if (pipelineUvn.equals("LATEST")) {
                    log = dbClient.fetchLatestLog(logKey);
                    if (log == null) {
                        return ResponseEntity.badRequest().body("No deployment log exists.");
                    }
                    pipelineUvn = log.getPipelineUniqueVersionNumber();
                }

                //TODO: This iteration is in enough places where it should be extracted as a dbClient method
                //Get most recent log (deployment or remediation) with desired pipeline UVN
                var logIncrement = 0;
                var logList = dbClient.fetchLogList(logKey, logIncrement);
                int idx = 0;
                while (idx < logList.size()) {
                    if (logList.get(idx).getPipelineUniqueVersionNumber().equals(pipelineUvn)) {
                        log = logList.get(idx);
                        break;
                    }
                    idx++;
                    if (idx == logList.size()) {
                        logIncrement++;
                        logList = dbClient.fetchLogList(logKey, logIncrement);
                        idx = 0;
                    }
                }

                if (log == null) {
                    status.markIncomplete();
                    continue;
                }
                if (log instanceof DeploymentLog && ((DeploymentLog) log).getStatus().equals(Log.LogStatus.PROGRESSING.name())) {
                    status.addProgressingStep(step);
                    continue;
                }
                if (log instanceof DeploymentLog && ((DeploymentLog) log).getStatus().equals(Log.LogStatus.CANCELLED.name())) {
                    status.markCancelled();
                    continue;
                }
                //Determines if the step is stable
                status.addLatestLog(log);
                //Determines if the step is complete
                if (log instanceof DeploymentLog) {
                    status.addLatestDeploymentLog((DeploymentLog) log, step);
                } else {
                    log = null;
                    while (idx < logList.size()) {
                        if (logList.get(idx).getPipelineUniqueVersionNumber().equals(pipelineUvn) && (logList.get(idx) instanceof DeploymentLog)) {
                            log = logList.get(idx);
                            status.addLatestDeploymentLog((DeploymentLog) log, step);
                            break;
                        }
                        idx++;
                        if (idx == logList.size()) {
                            logIncrement++;
                            logList = dbClient.fetchLogList(logKey, logIncrement);
                            idx = 0;
                        }
                    }
                    if (log == null) {
                        status.markIncomplete();
                        continue;
                    }
                }

                // was rollback deployment, need to get failure logs of initial deployment
                if (log.getUniqueVersionInstance() != 0) {
                    logIncrement = 0;
                    logList = dbClient.fetchLogList(logKey, logIncrement);
                    idx = 0;
                    while (idx < logList.size()) {
                        if (logList.get(idx).getUniqueVersionInstance() == 0 && (logList.get(idx) instanceof DeploymentLog)) {
                            status.addFailedDeploymentLog((DeploymentLog) logList.get(idx), step);
                            break;
                        }
                        idx++;
                        if (idx == logList.size()) {
                            logIncrement++;
                            logList = dbClient.fetchLogList(logKey, logIncrement);
                            idx = 0;
                        }
                    }
                }
            } catch (AtlasNonRetryableError e) {
                status.markIncomplete();
            }
        }
        return ResponseEntity.ok()
                .contentType(MediaType.APPLICATION_JSON)
                .body(schemaToResponsePayload(status));
    }

    @DeleteMapping(value = "{orgName}/{teamName}/pipelineRun/{pipelineName}")
    public ResponseEntity<Void> cancelLatestPipeline(@PathVariable("orgName") String orgName,
                                                     @PathVariable("teamName") String teamName,
                                                     @PathVariable("pipelineName") String pipelineName) {
        if (!requestSchemaValidator.checkAuthentication()) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
        }
        var latestUvn = "";
        var steps = dbClient.fetchStringList(DbKey.makeDbListOfStepsKey(orgName, teamName, pipelineName));
        for (var stepName : steps) {
            var key = DbKey.makeDbStepKey(orgName, teamName, pipelineName, stepName);
            var latestLog = dbClient.fetchLatestLog(key);
            if (latestUvn.isEmpty()) {
                //Step list is ordered, so if the very first log is nonexistant, it can't have deployed anywhere else
                if (latestLog == null) {
                    return ResponseEntity.ok().build();
                }
                latestUvn = latestLog.getPipelineUniqueVersionNumber();
            }
            if (latestLog == null || !latestLog.getPipelineUniqueVersionNumber().equals(latestUvn)) {
                var newCancelledLog = new DeploymentLog(latestUvn, Log.LogStatus.CANCELLED.name(), false, null, null);
                dbClient.insertValueInList(key, newCancelledLog);
            } else {
                latestLog.setStatus(Log.LogStatus.CANCELLED.name());
                dbClient.updateHeadInList(key, latestLog);
            }
        }
        return ResponseEntity.ok().build();
    }

    private String schemaToResponsePayload(Object schema) {
        try {
            return objectMapper.writeValueAsString(schema);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("Could not convert schema into response payload.", e);
        }
    }
}

File: workflowtrigger/api/pipeline_status_api.go
package api

import (
	"fmt"
	"github.com/gorilla/mux"
	"greenops.io/workflowtrigger/api/argoauthenticator"
	"greenops.io/workflowtrigger/api/reposerver"
	"greenops.io/workflowtrigger/db"
	"greenops.io/workflowtrigger/pipelinestatus"
	"greenops.io/workflowtrigger/util/auditlog"
	"greenops.io/workflowtrigger/util/serializer"
	"math"
	"net/http"
	"strconv"
)

const (
	stepNameField    string = "stepName"
	pipelineUvnField string = "pipelineUvn"
	countField       string = "count"
)

func getStepLogs(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	orgName := vars[orgNameField]
	teamName := vars[teamNameField]
	pipelineName := vars[pipelineNameField]
	stepName := vars[stepNameField]

	pipelineSchema := getPipeline(orgName, teamName, pipelineName)
	if !schemaValidator.ValidateSchemaAccess(orgName, teamName, pipelineSchema.GetGitRepoSchema().GitRepo, reposerver.RootCommit,
		string(argoauthenticator.GetAction), string(argoauthenticator.ApplicationResource)) {
		http.Error(w, "Not enough permissions", http.StatusForbidden)
		return
	}

	count, err := strconv.Atoi(vars[countField])
	if err != nil {
		http.Error(w, "Count variable could not be deserialized", http.StatusBadRequest)
		return
	}
	key := db.MakeDbStepKey(orgName, teamName, pipelineName, stepName)
	increments := int(math.Ceil(float64(db.LogIncrement / count)))
	var logList []auditlog.Log
	var fetchedLogList []auditlog.Log
	for idx := 0; idx < increments; idx++ {
		fetchedLogList = dbClient.FetchLogList(key, idx)
		if idx == increments-1 {
			difference := count - ((increments - 1) * db.LogIncrement)
			logList = append(logList, fetchedLogList[0:int(math.Min(float64(difference), float64(len(fetchedLogList))))]...)
		} else {
			logList = append(logList, fetchedLogList...)
		}
	}
	w.Header().Set("Content-Type", "application/json")
	w.Write([]byte(serializer.Serialize(logList)))
}

func getPipelineUvns(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	orgName := vars[orgNameField]
	teamName := vars[teamNameField]
	pipelineName := vars[pipelineNameField]

	pipelineSchema := getPipeline(orgName, teamName, pipelineName)
	if !schemaValidator.ValidateSchemaAccess(orgName, teamName, pipelineSchema.GetGitRepoSchema().GitRepo, reposerver.RootCommit,
		string(argoauthenticator.GetAction), string(argoauthenticator.ApplicationResource)) {
		http.Error(w, "Not enough permissions", http.StatusForbidden)
		return
	}

	count, err := strconv.Atoi(vars[countField])
	if err != nil {
		http.Error(w, "Count variable could not be deserialized", http.StatusBadRequest)
		return
	}
	key := db.MakeDbPipelineInfoKey(orgName, teamName, pipelineName)
	increments := int(math.Ceil(float64(db.LogIncrement / count)))
	var pipelineUvnList []string
	var fetchedPipelineUvnList []string
	for idx := 0; idx < increments; idx++ {
		for _, pipelineInfo := range dbClient.FetchPipelineInfoList(key, idx) {
			fetchedPipelineUvnList = append(fetchedPipelineUvnList, pipelineInfo.PipelineUvn)
		}
		if idx == increments-1 {
			difference := count - ((increments - 1) * db.LogIncrement)
			pipelineUvnList = append(pipelineUvnList, fetchedPipelineUvnList[0:int(math.Min(float64(difference), float64(len(fetchedPipelineUvnList))))]...)
		} else {
			pipelineUvnList = append(pipelineUvnList, fetchedPipelineUvnList...)
		}
	}
	w.Header().Set("Content-Type", "application/json")
	w.Write([]byte(serializer.Serialize(pipelineUvnList)))
}

func getPipelineStatus(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	orgName := vars[orgNameField]
	teamName := vars[teamNameField]
	pipelineName := vars[pipelineNameField]
	pipelineUvn := vars[pipelineUvnField]

	pipelineSchema := getPipeline(orgName, teamName, pipelineName)
	if !schemaValidator.ValidateSchemaAccess(orgName, teamName, pipelineSchema.GetGitRepoSchema().GitRepo, reposerver.RootCommit,
		string(argoauthenticator.GetAction), string(argoauthenticator.ApplicationResource)) {
		http.Error(w, "Not enough permissions", http.StatusForbidden)
		return
	}

	status := pipelinestatus.New()
	//Get pipeline UVN if not specified
	pipelineInfoKey := db.MakeDbPipelineInfoKey(orgName, teamName, pipelineName)
	var pipelineInfo auditlog.PipelineInfo
	if pipelineUvn == "LATEST" {
		pipelineInfo = dbClient.FetchLatestPipelineInfo(pipelineInfoKey)
		pipelineUvn = pipelineInfo.PipelineUvn
	} else {
		logIncrement := 0
		pipelineInfoList := dbClient.FetchPipelineInfoList(pipelineInfoKey, logIncrement)
		idx := 0
		for idx < len(pipelineInfoList) {
			if pipelineInfoList[idx].PipelineUvn == pipelineUvn {
				pipelineInfo = pipelineInfoList[idx]
				break
			}
			idx++
			if idx == len(pipelineInfoList) {
				logIncrement++
				pipelineInfoList = dbClient.FetchPipelineInfoList(pipelineInfoKey, logIncrement)
				idx = 0
			}
		}
	}
	if pipelineInfo.PipelineUvn == "" && len(pipelineInfo.Errors) == 0 {
		http.Error(w, "No pipeline runs exist with the requested UVN", http.StatusBadRequest)
		return
	}
	steps := pipelineInfo.StepList
	for _, step := range steps {
		//Get pipeline UVN if not specified
		logKey := db.MakeDbStepKey(orgName, teamName, pipelineName, step)
		var log auditlog.Log

		//TODO: This iteration is in enough places where it should be extracted as a dbClient method
		//Get most recent log (deployment or remediation) with desired pipeline UVN
		logIncrement := 0
		logList := dbClient.FetchLogList(logKey, logIncrement)
		idx := 0
		for idx < len(logList) {
			if logList[idx].GetPipelineUniqueVersionNumber() == pipelineUvn {
				log = logList[idx]
				break
			}
			idx++
			if idx == len(logList) {
				logIncrement++
				logList = dbClient.FetchLogList(logKey, logIncrement)
				idx = 0
			}
		}

		if log == nil {
			status.MarkIncomplete()
			continue
		}
		if tempDeploymentLog, ok := log.(*auditlog.DeploymentLog); ok {
			if tempDeploymentLog.GetStatus() == auditlog.Progressing {
				status.AddProgressingStep(step)
				continue
			}
		}
		if tempDeploymentLog, ok := log.(*auditlog.DeploymentLog); ok {
			if tempDeploymentLog.GetStatus() == auditlog.Cancelled {
				status.MarkCancelled()
				if tempDeploymentLog.BrokenTest != "" {
					status.AddFailedDeploymentLog(*tempDeploymentLog, step)
				}
				continue
			}
		}
		//Determines if the step is stable
		status.AddLatestLog(log)
		//Determines if the step is complete
		if tempDeploymentLog, ok := log.(*auditlog.DeploymentLog); ok {
			status.AddLatestDeploymentLog(*tempDeploymentLog, step)
		} else {
			log = nil
			for idx < len(logList) {
				if logList[idx].GetPipelineUniqueVersionNumber() == pipelineUvn {
					if _, ok = logList[idx].(*auditlog.DeploymentLog); ok {
						log = logList[idx]
						status.AddLatestDeploymentLog(*(log.(*auditlog.DeploymentLog)), step)
						break
					}
				}
				idx++
				if idx == len(logList) {
					logIncrement++
					logList = dbClient.FetchLogList(logKey, logIncrement)
					idx = 0
				}
			}
			if log == nil {
				status.MarkIncomplete()
				continue
			}
		}

		// was rollback deployment, need to get failure logs of initial deployment
		if log.GetUniqueVersionInstance() != 0 {
			logIncrement = 0
			logList = dbClient.FetchLogList(logKey, logIncrement)
			idx = 0
			for idx < len(logList) {
				if logList[idx].GetUniqueVersionInstance() == 0 {
					if tempDeploymentLog, ok := logList[idx].(*auditlog.DeploymentLog); ok {
						status.AddFailedDeploymentLog(*tempDeploymentLog, step)
						break
					}
				}
				idx++
				if idx == len(logList) {
					logIncrement++
					logList = dbClient.FetchLogList(logKey, logIncrement)
					idx = 0
				}
			}
		}
	}
	//Get additional "floating" errors. Largely related to processing
	if !(pipelineInfo.PipelineUvn == "" && len(pipelineInfo.Errors) == 0) {
		for idx, err := range pipelineInfo.Errors {
			status.AddFailedDeploymentLog(auditlog.DeploymentLog{
				DeploymentComplete: false,
				BrokenTest:         fmt.Sprintf("Processing error %d", idx),
				BrokenTestLog:      err,
			}, "Unknown")
		}
	}
	w.Header().Set("Content-Type", "application/json")
	w.Write([]byte(serializer.Serialize(status)))
}

func cancelLatestPipeline(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	orgName := vars[orgNameField]
	teamName := vars[teamNameField]
	pipelineName := vars[pipelineNameField]

	pipelineSchema := getPipeline(orgName, teamName, pipelineName)
	if !schemaValidator.ValidateSchemaAccess(orgName, teamName, pipelineSchema.GetGitRepoSchema().GitRepo, reposerver.RootCommit,
		string(argoauthenticator.SyncAction), string(argoauthenticator.ApplicationResource),
		string(argoauthenticator.SyncAction), string(argoauthenticator.ClusterResource)) {
		http.Error(w, "Not enough permissions", http.StatusForbidden)
		return
	}

	latestUvn := ""
	steps := dbClient.FetchLatestPipelineInfo(db.MakeDbPipelineInfoKey(orgName, teamName, pipelineName)).StepList
	for _, stepName := range steps {
		key := db.MakeDbStepKey(orgName, teamName, pipelineName, stepName)
		latestLog := dbClient.FetchLatestLog(key)
		if len(latestUvn) == 0 {
			//Step list is ordered, so if the very first log is nonexistant, it can't have deployed anywhere else
			if latestLog == nil {
				w.WriteHeader(http.StatusOK)
				return
			}
			latestUvn = latestLog.GetPipelineUniqueVersionNumber()
		}
		if latestLog == nil || latestLog.GetPipelineUniqueVersionNumber() != latestUvn {
			newCancelledLog := auditlog.InitBlankDeploymentLog(latestUvn, auditlog.Cancelled, false, "", "")
			dbClient.InsertValueInList(key, newCancelledLog)
		} else {
			latestLog.SetStatus(auditlog.Cancelled)
			dbClient.UpdateHeadInList(key, latestLog)
		}
	}
	w.WriteHeader(http.StatusOK)
	return
}

func InitStatusEndpoints(r *mux.Router) {
	r.HandleFunc("/status/{orgName}/{teamName}/pipeline/{pipelineName}/history/{count}", getPipelineUvns).Methods("GET")
	r.HandleFunc("/status/{orgName}/{teamName}/pipeline/{pipelineName}/step/{stepName}/{count}", getStepLogs).Methods("GET")
	r.HandleFunc("/status/{orgName}/{teamName}/pipeline/{pipelineName}/{pipelineUvn}", getPipelineStatus).Methods("GET")
	r.HandleFunc("/status/{orgName}/{teamName}/pipelineRun/{pipelineName}", cancelLatestPipeline).Methods("DELETE")
}
